nikola/plugin_manager.py

```diff
@@ -0,0 +1,234 @@
+                if name_parts[-1] == "__init__.py":
+                    name_parts.pop(-1)
+                elif name_parts[-1].endswith(".py"):
+                    name_parts[-1] = name_parts[-1][:-3]
+                full_module_name = ".".join(name_parts)
+            except ValueError:
+                pass
+
+            try:
+                spec = importlib.util.spec_from_file_location(full_module_name, py_file_location)
+                module_object = importlib.util.module_from_spec(spec)
+                if full_module_name not in sys.modules:
+                    sys.modules[full_module_name] = module_object
+                spec.loader.exec_module(module_object)
+            except Exception:
+                self.logger.exception(f"{plugin_id} threw an exception while loading")
+                continue
+
+            plugin_classes = [
+                c
+                for c in vars(module_object).values()
+                if isinstance(c, type) and issubclass(c, BasePlugin) and c not in CATEGORY_TYPES
+            ]
+            if len(plugin_classes) == 0:
+                self.logger.warning(f"{plugin_id} does not have any plugin classes")
+                continue
+            elif len(plugin_classes) > 1:
+                self.logger.warning(f"{plugin_id} has multiple plugin classes; this is not supported - skipping")
+                continue
+            try:
+                plugin_object = plugin_classes[0]()
+            except Exception:
+                self.logger.exception(f"{plugin_id} threw an exception while creating an instance")
+                continue
+            self.logger.debug(f"Loaded {plugin_id}")
+            info = PluginInfo(
+                name=name,
+                description=candidate.description,
+                plugin_id=candidate.plugin_id,
+                category=candidate.category,
+                compiler=candidate.compiler,
+                source_dir=source_dir,
+                module_name=module_name,
+                module_object=module_object,
+                plugin_object=plugin_object,
+            )
+            self.plugins.append(info)
+
+        self._plugins_by_category = {category: [] for category in CATEGORY_NAMES}
+        for plugin_info in self.plugins:
+            self._plugins_by_category[plugin_info.category].append(plugin_info)
+
+    def get_plugins_of_category(self, category: str) -> List[PluginInfo]:
+        """Get loaded plugins of a given category."""
+        return self._plugins_by_category.get(category, [])
+
+    def get_plugin_by_name(self, name: str, category: Optional[str] = None) -> Optional[PluginInfo]:
+        """Get a loaded plugin by name and optionally by category. Returns None if no such plugin is loaded."""
+        for p in self.plugins:
+            if p.name == name and (category is None or p.category == category):
+                return p
+
+    # Aliases for Yapsy compatibility
```

Comment:
Well done!

How about deprecating these, so these can be removed with Nikola 3.0.0?

Reply:
It's a small issue, but I'm for cleaning stuff like this up, long-term.

I have some code to log a warning for each source code place that uses these, the first time they do so.

Technically, this could have been done via the [warning module](https://docs.python.org/release/3.8.18/library/warnings.html). But then, people would not see it unless they specifically look for it. Instead, I think it better to just rolled it manually via `self.logger`.

Depending whether this is what you want to do from a project governance point of view (as I recommend), I can contribute that code. Would you like my pull request targeting the `new-plugin-manager` branch?

Reply:
Later: As I had the code already, I just prepared that PR #3730. I won't be offended if you don't want this functionality and close it. If that PR is accepted, it's work will be merged into this PR, not into `master`. I don't mind if, in the end, my work and the other work in this branch are all squashed into one commit and I'm no longer mentioned as an author in the git history.

Reply:
I’ve rejected that PR, since it was adding a lot of unnecessary complexity, while a simple log message would suffice (I’ve added it).

Reply:
Ok.

Reply:
Two plugins in getnikola/plugins are using the old names BTW.

Reply:
@felixfontein Yeah, guess we’ll need to update them at some point.